### PR TITLE
Added example utility using OOPS and IODA

### PR DESCRIPTION
**Pull request by CoryMartin-NOAA**
This PR adds a very simple utility that will read in a IODA file, and for a specified group/variable, compute the mean and print it out. Nothing too fancy but is a nice C++ OOPS example.

### CODE REVIEW

**Comment by CoryMartin-NOAA** [comment]
@guillaumevernieres make sure I didn't screw up the merge conflicts

**Review by RussTreadon-NOAA — COMMENTED** [review]
Thank you for adding this oops example.  This is a handy reference.   Changes look good to my inexperienced eyes.  I have not installed `feature/oops_example` on Hera or Orion and run ctests to ensure everything works a planned.  Adding a CI label to this PR will do this for us.

File: utils/ioda_example/gdas_meanioda.h
Comment (guillaumevernieres) [inline utils/ioda_example/gdas_meanioda.h:36]: how about adding an empty line after all the sequential steps? It would make the code a bit nicer to read.

**Review by guillaumevernieres — DISMISSED** [review]
Thanks for doing this @CoryMartin-NOAA . I have a minor cosmetic comment that you can decide if you want to address or not. Ready to merge, pending CI results. Code norm to be added in a separate PR.

**Comment by RussTreadon-NOAA** [comment]
**Orion build failure**

Clone `feature/oops_example` in working copy of g-w `develop`.   `build_gdas.sh` fails with the following message in `/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/logs/build_gdas.log`

```
-- Configuring done
CMake Error at /work/noaa/epic-ps/role-epic-ps/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2022.0.2/ecbuild-3.7.2-zgkban6/share/ecbuild/cmake/ecbuild_add_executable.cmake:209 (add_executable):
  Cannot find source file:

    gdas_incr_handler.cc
Call Stack (most recent call first):
  gdas-utils/soca/CMakeLists.txt:2 (ecbuild_add_executable)


CMake Error at /work/noaa/epic-ps/role-epic-ps/spack-stack/spack-stack-1.3.1/envs/unified-env/install/intel/2022.0.2/ecbuild-3.7.2-zgkban6/share/ecbuild/cmake/ecbuild_add_executable.cmake:209 (add_executable):
  No SOURCES given to target: gdas_incr_handler.x
Call Stack (most recent call first):
  gdas-utils/soca/CMakeLists.txt:2 (ecbuild_add_executable)


CMake Generate step failed.  Build files cannot be regenerated correctly.
```

**Comment by CoryMartin-NOAA** [comment]
Yeah it should be fixed now, let me re trigger the Hera CI

**Comment by RussTreadon-NOAA** [comment]
> Yeah it should be fixed now, let me re trigger the Hera CI

Updates pulled into Orion working copy and rerunning `build_gdas.sh`.

File: utils/ioda_example/gdas_meanioda.h
Comment (CoryMartin-NOAA) [reply utils/ioda_example/gdas_meanioda.h:36]: Fixed at https://github.com/NOAA-EMC/GDASApp/pull/553/commits/8bb6a361e73a27883e718a65f1e977bf99632224

**Comment by RussTreadon-NOAA** [comment]
**Orion test**

Successful build of `feature/oops_example` at 8bb6a36.   `test_gdasapp` executed from within g-w with the following results

```
(gdasapp) Orion-login-4:/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/build$ ctest -R test_gdasapp
Test project /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/build
      Start 1243: test_gdasapp_util_ioda_example
 1/54 Test #1243: test_gdasapp_util_ioda_example ........................***Failed   12.73 sec
      Start 1559: test_gdasapp_check_python_norms
 2/54 Test #1559: test_gdasapp_check_python_norms .......................   Passed    1.11 sec
      Start 1560: test_gdasapp_check_yaml_keys
 3/54 Test #1560: test_gdasapp_check_yaml_keys ..........................   Passed    0.56 sec
      Start 1561: test_gdasapp_jedi_increment_to_fv3
 4/54 Test #1561: test_gdasapp_jedi_increment_to_fv3 ....................   Passed    4.24 sec
      Start 1562: test_gdasapp_genYAML_prep
 5/54 Test #1562: test_gdasapp_genYAML_prep .............................   Passed    0.06 sec
      Start 1563: test_gdasapp_genYAML_run
 6/54 Test #1563: test_gdasapp_genYAML_run ..............................   Passed    9.64 sec
      Start 1564: test_gdasapp_genYAML_land_prep
 7/54 Test #1564: test_gdasapp_genYAML_land_prep ........................   Passed    0.02 sec
      Start 1565: test_gdasapp_genYAML_land_run
 8/54 Test #1565: test_gdasapp_genYAML_land_run .........................   Passed    0.55 sec
      Start 1566: test_gdasapp_genYAML_aero_prep
 9/54 Test #1566: test_gdasapp_genYAML_aero_prep ........................   Passed    0.02 sec
      Start 1567: test_gdasapp_genYAML_aero_run
10/54 Test #1567: test_gdasapp_genYAML_aero_run .........................   Passed    0.54 sec
      Start 1568: test_gdasapp_convert_ewok_yaml
11/54 Test #1568: test_gdasapp_convert_ewok_yaml ........................   Passed    0.16 sec
      Start 1569: test_gdasapp_setup_cycled_exp
12/54 Test #1569: test_gdasapp_setup_cycled_exp .........................   Passed    0.40 sec
      Start 1570: test_gdasapp_convert_bufr_temp_dbuoy
13/54 Test #1570: test_gdasapp_convert_bufr_temp_dbuoy ..................   Passed    3.13 sec
      Start 1571: test_gdasapp_convert_bufr_salt_dbuoy
14/54 Test #1571: test_gdasapp_convert_bufr_salt_dbuoy ..................   Passed    0.19 sec
      Start 1572: test_gdasapp_convert_bufr_temp_mbuoyb
15/54 Test #1572: test_gdasapp_convert_bufr_temp_mbuoyb .................   Passed    0.18 sec
      Start 1573: test_gdasapp_convert_bufr_salt_mbuoyb
16/54 Test #1573: test_gdasapp_convert_bufr_salt_mbuoyb .................   Passed    0.18 sec
      Start 1574: test_gdasapp_convert_bufr_tesacprof
17/54 Test #1574: test_gdasapp_convert_bufr_tesacprof ...................   Passed    0.21 sec
      Start 1575: test_gdasapp_convert_bufr_trkobprof
18/54 Test #1575: test_gdasapp_convert_bufr_trkobprof ...................   Passed    0.18 sec
      Start 1576: test_gdasapp_convert_bufr_sfcships
19/54 Test #1576: test_gdasapp_convert_bufr_sfcships ....................   Passed    0.19 sec
      Start 1577: test_gdasapp_convert_bufr_sfcshipsu
20/54 Test #1577: test_gdasapp_convert_bufr_sfcshipsu ...................   Passed    0.18 sec
      Start 1578: test_gdasapp_soca_obsdb
21/54 Test #1578: test_gdasapp_soca_obsdb ...............................   Passed    0.83 sec
      Start 1579: test_gdasapp_soca_nsst_increment_to_mom6
22/54 Test #1579: test_gdasapp_soca_nsst_increment_to_mom6 ..............   Passed   33.11 sec
      Start 1580: test_gdasapp_soca_prep
23/54 Test #1580: test_gdasapp_soca_prep ................................   Passed    3.90 sec
      Start 1581: test_gdasapp_soca_concatioda
24/54 Test #1581: test_gdasapp_soca_concatioda ..........................   Passed    2.62 sec
      Start 1582: test_gdasapp_soca_run_clean
25/54 Test #1582: test_gdasapp_soca_run_clean ...........................   Passed    0.19 sec
      Start 1583: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP
26/54 Test #1583: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_PREP ....   Passed  170.65 sec
      Start 1584: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT
27/54 Test #1584: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_BMAT ....   Passed  106.19 sec
      Start 1585: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN
28/54 Test #1585: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_RUN .....   Passed   42.19 sec
      Start 1586: test_gdasapp_soca_copy_scratch
29/54 Test #1586: test_gdasapp_soca_copy_scratch ........................   Passed    6.06 sec
      Start 1587: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT
30/54 Test #1587: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_CHKPT ...   Passed   42.18 sec
      Start 1588: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST
31/54 Test #1588: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_POST ....   Passed   42.25 sec
      Start 1589: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY
32/54 Test #1589: test_gdasapp_soca_JGDAS_GLOBAL_OCEAN_ANALYSIS_VRFY ....   Passed  203.75 sec
      Start 1590: test_gdasapp_soca_socahybridweights
33/54 Test #1590: test_gdasapp_soca_socahybridweights ...................   Passed   42.16 sec
      Start 1591: test_gdasapp_soca_incr_handler
34/54 Test #1591: test_gdasapp_soca_incr_handler ........................   Passed   42.16 sec
      Start 1592: test_gdasapp_land_create_ens
35/54 Test #1592: test_gdasapp_land_create_ens ..........................   Passed    1.16 sec
      Start 1593: test_gdasapp_land_imsproc
36/54 Test #1593: test_gdasapp_land_imsproc .............................   Passed    8.68 sec
      Start 1594: test_gdasapp_land_apply_jediincr
37/54 Test #1594: test_gdasapp_land_apply_jediincr ......................   Passed    5.54 sec
      Start 1595: test_gdasapp_land_letkfoi_snowda
38/54 Test #1595: test_gdasapp_land_letkfoi_snowda ......................   Passed   15.24 sec
      Start 1596: test_gdasapp_convert_bufr_adpsfc_snow
Could not find executable /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
Looked in the following places:
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Release/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Release/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Debug/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Debug/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/MinSizeRel/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/MinSizeRel/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/RelWithDebInfo/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/RelWithDebInfo/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Deployment/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Deployment/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Development/test_bufr2ioda.sh
/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Development/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Release/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Release/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Debug/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Debug/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/MinSizeRel/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/MinSizeRel/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/RelWithDebInfo/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/RelWithDebInfo/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Deployment/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Deployment/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Development/test_bufr2ioda.sh
work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/Development/test_bufr2ioda.sh
Unable to find executable: /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/test/land/test_bufr2ioda.sh
39/54 Test #1596: test_gdasapp_convert_bufr_adpsfc_snow .................***Not Run   0.00 sec
      Start 1597: test_gdasapp_convert_bufr_adpsfc
40/54 Test #1597: test_gdasapp_convert_bufr_adpsfc ......................   Passed    7.22 sec
      Start 1598: test_gdasapp_convert_gsi_satbias
41/54 Test #1598: test_gdasapp_convert_gsi_satbias ......................   Passed    4.24 sec
      Start 1599: test_gdasapp_store_gsi_satbias
42/54 Test #1599: test_gdasapp_store_gsi_satbias ........................   Passed    0.68 sec
      Start 1600: test_gdasapp_run_jedi_exe_3dhofx
43/54 Test #1600: test_gdasapp_run_jedi_exe_3dhofx ......................   Passed   17.25 sec
      Start 1601: test_gdasapp_run_jedi_exe_3dvar
44/54 Test #1601: test_gdasapp_run_jedi_exe_3dvar .......................   Passed   10.91 sec
      Start 1602: test_gdasapp_run_jedi_exe_3denvar
45/54 Test #1602: test_gdasapp_run_jedi_exe_3denvar .....................   Passed   11.47 sec
      Start 1603: test_gdasapp_run_jedi_exe_letkf
46/54 Test #1603: test_gdasapp_run_jedi_exe_letkf .......................   Passed    6.83 sec
      Start 1604: test_gdasapp_setup_atm_cycled_exp
47/54 Test #1604: test_gdasapp_setup_atm_cycled_exp .....................   Passed    0.64 sec
      Start 1605: test_gdasapp_atm_jjob_var_init
48/54 Test #1605: test_gdasapp_atm_jjob_var_init ........................***Failed   43.16 sec
      Start 1606: test_gdasapp_atm_jjob_var_run
49/54 Test #1606: test_gdasapp_atm_jjob_var_run .........................***Failed   42.10 sec
      Start 1607: test_gdasapp_atm_jjob_var_final
50/54 Test #1607: test_gdasapp_atm_jjob_var_final .......................***Failed   10.09 sec
      Start 1608: test_gdasapp_atm_jjob_ens_init
51/54 Test #1608: test_gdasapp_atm_jjob_ens_init ........................   Passed   43.62 sec
      Start 1609: test_gdasapp_atm_jjob_ens_run
52/54 Test #1609: test_gdasapp_atm_jjob_ens_run .........................***Failed   74.14 sec
      Start 1610: test_gdasapp_atm_jjob_ens_final
53/54 Test #1610: test_gdasapp_atm_jjob_ens_final .......................***Failed   42.11 sec
      Start 1611: test_gdasapp_aero_gen_3dvar_yaml
54/54 Test #1611: test_gdasapp_aero_gen_3dvar_yaml ......................   Passed    0.24 sec

87% tests passed, 7 tests failed out of 54

Label Time Summary:
gdas-utils    =  12.73 sec*proc (1 test)
script        =  12.73 sec*proc (1 test)

Total Test time (real) = 1121.46 sec

The following tests FAILED:
        1243 - test_gdasapp_util_ioda_example (Failed)
        1596 - test_gdasapp_convert_bufr_adpsfc_snow (Not Run)
        1605 - test_gdasapp_atm_jjob_var_init (Failed)
        1606 - test_gdasapp_atm_jjob_var_run (Failed)
        1607 - test_gdasapp_atm_jjob_var_final (Failed)
        1609 - test_gdasapp_atm_jjob_ens_run (Failed)
        1610 - test_gdasapp_atm_jjob_ens_final (Failed)
Errors while running CTest
Output from these tests are in: /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/build/Testing/Temporary/LastTest.log
Use "--rerun-failed --output-on-failure" to re-run the failed cases verbosely.
```

The `atm_jjob` failures appear to be related to CRTM fix files.  An investigation is currently underway.

**Comment by RussTreadon-NOAA** [comment]
`atm_jjob_var_init` failure due to working copy of g-w **_not_** including g-w PR [#1773](https://github.com/NOAA-EMC/global-workflow/pull/1773).  Update working copy of g-w and rerun.  `atm_jjob_var_init` passed.   Variational and ensemble run jobs still fail.  Traceback below

```
1: Exception:   source_column:  0
1:      source_filename:        /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/ioda/src/engines/ioda/include/ioda/Variables/Variable.h
1:      source_function:        Variable_Implementation ioda::detail::Variable_Base<Variable_Implementation>::read(gsl::span<DataType>, const ioda::Selection &, const ioda::Selection &) const [with DataType = int; Marshaller = ioda::detail::Object_Accessor_Regular<int, int>; TypeWrapper = ioda::Types::GetType_Wrapper<int, 0>; Variable_Implementation = ioda::Variable]
1:      source_line:    539
1:
1: Exception: oops::Variational<FV3JEDI, UFO and IODA observations> terminating...
```

This error is unrelated to this PR.

**Comment by RussTreadon-NOAA** [comment]
Comment out amsua_n19 `obs space` section of `gdas.t18z.atmvar.yaml` used by `atm_jjob_var_run`.   With this change, the `atm_jjob_var_run` ctest passed ... so maybe still dealing with a crtm issue.
```
(gdasapp) Orion-login-4:/work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/build$ ctest -R test_gdasapp_atm_jjob_var_run
Test project /work2/noaa/da/rtreadon/git/global-workflow/pr553/sorc/gdas.cd/build
    Start 1606: test_gdasapp_atm_jjob_var_run
1/1 Test #1606: test_gdasapp_atm_jjob_var_run ....   Passed  203.61 sec

The following tests passed:
        test_gdasapp_atm_jjob_var_run

100% tests passed, 0 tests failed out of 1

Total Test time (real) = 204.10 sec
```

`atm_jjob_ens_run` includes an amsua_n19 `obs_space` section.   The atm_jjob `final` jobs fail because expected output from the `run` jobs is not found.

**Comment by RussTreadon-NOAA** [comment]
Last update regarding `test_gdasapp_atm_jjob_var_run` failure.   Comment out all amsua_n19 obs filters, obsdataout, obs bias.  `fv3jedi_var.x` still aborts as noted above.  This suggests that we need to update the input amsua_n19 obs file.  Not sure what needs to be changed.   Tagging @emilyhcliu for awareness.

**Comment by CoryMartin-NOAA** [comment]
On Hera, compiling is failing when it gets to my new code...looks like it is a C++17 issue, I need to figure out what the issue is on Hera that work on Orion and my Linux desktop with Singularity

**Comment by CoryMartin-NOAA** [comment]
C++17 on Hera issue fixed by using a function that is 'older'. Not sure why that didn't build, I thought the compiler was new enough on Hera (this is concerning but we will cross that bridge later). Will try to loo into the failing tests some more now.

**Comment by CoryMartin-NOAA** [comment]
@RussTreadon-NOAA I bet this is the culprit: https://github.com/JCSDA-internal/ufo/commit/92f78421da2713bb256f8f758381381060bad6a2

`sensorScanPosition` is float in our files, it needs to be int.

**Comment by CoryMartin-NOAA** [comment]
Apparently the CI testing does not download the JCSDA test data. @guillaumevernieres @RussTreadon-NOAA should it do that? or should I just manually include/download a single IODA file for this test? I could include one in our testdata tar file.

**Comment by RussTreadon-NOAA** [comment]
> Apparently the CI testing does not download the JCSDA test data. @guillaumevernieres @RussTreadon-NOAA should it do that? or should I just manually include/download a single IODA file for this test? I could include one in our testdata tar file.

Which approach is easier to implement now and maintain moving forward?

**Comment by CoryMartin-NOAA** [comment]
Probably cloning the jcsda test data, but it might be overkill. I'm happy to do that if we think that's the easiest path forward.

**Comment by guillaumevernieres** [comment]
> Probably cloning the jcsda test data, but it might be overkill. I'm happy to do that if we think that's the easiest path forward.

@CoryMartin-NOAA , can you run your test with one of the soca ioda files? soca still uses git-lfs to store binaries.

**Comment by CoryMartin-NOAA** [comment]
I went with @guillaumevernieres 's suggestion, it now uses one of the SOCA test files that have brightnessTemperature as a variable.

**Comment by RussTreadon-NOAA** [comment]
`test_gdasapp_util_ioda_example` _Passed_ on Orion following update at [`55f1b6c`](https://github.com/NOAA-EMC/GDASApp/pull/553/commits/55f1b6c281445f6e3603ddd902566fcc36541f5c).

**Comment by CoryMartin-NOAA** [comment]
The failures are unrelated to this PR it seems, but I defer to @RussTreadon-NOAA and @guillaumevernieres if we want to try to fix them before approving/merging this.

**Comment by RussTreadon-NOAA** [comment]
issue #554 has been opened to report the `var_run` and `ens_run` failures.  A separate PR can be opened to resolve whatever isn't right.  No need to hold up this PR for what seems to be an ioda or ufo problem.

**Review by guillaumevernieres — APPROVED** [review]
I've been voting for ignoring the atmos ctest for the last few pr's! Looks ready to merge. Thanks again @CoryMartin-NOAA .

**Pull request merged**
